accord-core/src/main/java/accord/primitives/Routables.java

```diff
@@ -55,22 +81,61 @@
     /**
      * Perform {@link SortedArrays#exponentialSearch} from {@code thisIndex} looking for {@code find} with behaviour of {@code search}
      */
-    int findNext(int thisIndex, K find, SortedArrays.Search search);
+    int findNext(int thisIndex, RoutableKey find, SortedArrays.Search search);
 
-    Routable.Domain kindOfContents();
+    Domain domain();
+
+    @Inline
+    static <Param, Output> Output switchOnDomain(Routables<?, ?> inputs, Param param, BiFunction<AbstractKeys<?, ?>, Param, Output> ifKeys, BiFunction<AbstractRanges<?>, Param, Output> ifRanges)
+    {
+        switch (inputs.domain())
+        {
+            default: throw new AssertionError();
+            case Key: return ifKeys.apply((AbstractKeys<?, ?>) inputs, param);
+            case Range: return ifRanges.apply((AbstractRanges<?>) inputs, param);
+        }
+    }
+
+    @Inline
+    static <Param, Output> Output switchOnDomain(Domain domain, Param param, Function<Param, Output> ifKeys, Function<Param, Output> ifRanges)
+    {
+        switch (domain)
+        {
+            default: throw new AssertionError();
+            case Key: return ifKeys.apply(param);
+            case Range: return ifRanges.apply(param);
+        }
+    }
 
     @Inline
     static <Input extends Routable, T> T foldl(Routables<Input, ?> inputs, AbstractRanges<?> matching, IndexedFold<? super Input, T> fold, T initialValue)
     {
         return Helper.foldl(Routables::findNextIntersection, Helper::findLimit, inputs, matching, fold, initialValue);
     }
 
+    @Inline
+    static <T> T foldlMinimal(Seekables<?, ?> inputs, AbstractRanges<?> matching, IndexedFold<? super Seekable, T> fold, T initialValue)
+    {
+        return Helper.foldlMinimal(inputs, matching, fold, initialValue);
```

Comment:
Have added javadoc

Reply:
3 of the folds that are marked unused in Accord also seem to be unused in Cassandra. For range folding documenting what an indexed fold is would be helpful.

One of the range folds also looks like it might be unused.

`foldlMissing` what is the missing aspect?

Reply:
I would prefer not to prune the fold functions for now, as the specific ones in use vary a great deal between patches, so we may end up just re-implementing and reviewing them.

I documented foldMissing at the public API, but I can copy the documentation over to the Helper methods too if you like.

Reply:
Ah, got it, it's non-intersection.

Reply:
Could call it foldComplement if that’s clearer? Or some other equivalent term